kong/dns/client.lua

```diff
@@ -0,0 +1,701 @@
+  }
+
+  local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+    ipc = ipc,
+    neg_ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    lru_size = opts.cache_size or 10000,
+    shm_locks = ngx.shared.kong_locks and "kong_locks",
+    resty_lock_opts = resty_lock_opts,
+  })
+
+  if not cache then
+    return nil, "could not create mlcache: " .. err
+  end
+
+  if opts.cache_purge then
+    cache:purge(true)
+  end
+
+  -- parse hosts
+  local hosts, hosts_cache = init_hosts(cache, opts.hosts)
+
+  return setmetatable({
+    cache = cache,
+    stats = {},
+    hosts = hosts,
+    r_opts = r_opts,
+    resolv = opts._resolv or resolv,
+    valid_ttl = opts.valid_ttl,
+    error_ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl = opts.stale_ttl or DEFAULT_STALE_TTL,
+    enable_srv = enable_srv,
+    enable_ipv4 = enable_ipv4,
+    enable_ipv6 = enable_ipv6,
+    hosts_cache = hosts_cache,
+
+    -- TODO: Make the table readonly. But if `string.buffer.encode/decode` and
+    -- `pl.tablex.readonly` are called on it, it will become empty table.
+    --
+    -- quickly accessible constant empty answers
+    EMPTY_ANSWERS = {
+      errcode = EMPTY_RECORD_ERROR_CODE,
+      errstr = EMPTY_RECORD_ERROR_MESSAGE,
+      ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    },
+  }, MT)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = self.error_ttl
+    return answers
+  end
+
+  local processed_answers = {}
+
+  -- 0xffffffff for maximum TTL value
+  local ttl = math_min(self.valid_ttl or 0xffffffff, 0xffffffff)
+
+  for _, answer in ipairs(answers) do
+    answer.name = string_lower(answer.name)
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+
+    local answer_type = answer.type
+
+    if answer_type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer_type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+
+      elseif answer_type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      table_insert(processed_answers, answer)
+    end
+  end
+
+  if table_isempty(processed_answers) then
+    log(DEBUG, PREFIX, "processed ans:empty")
+    return self.EMPTY_ANSWERS
+  end
+
+  log(DEBUG, PREFIX, "processed ans:", #processed_answers)
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+
+  stats_init_name(self.stats, key)
+  stats_increment(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local start = now()
+
+  local answers, err = r:query(name, { additional_section = true, qtype = qtype })
```

Comment:
Just a note that this is not something we want, see the thread in Slack.

Reply:
yeah, I'll investigate and see how to fix it. I guess we just need to remove `addtional_setion=true` here.

Reply:
resolved!